processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnPartSerde.java

```diff
@@ -19,66 +19,226 @@
 
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Suppliers;
 import com.google.common.primitives.Ints;
 import com.metamx.collections.bitmap.ImmutableBitmap;
 import com.metamx.collections.spatial.ImmutableRTree;
 import com.metamx.common.IAE;
+import com.metamx.common.Pair;
 import io.druid.segment.column.ColumnBuilder;
 import io.druid.segment.column.ColumnConfig;
 import io.druid.segment.column.ValueType;
 import io.druid.segment.data.BitmapSerde;
 import io.druid.segment.data.BitmapSerdeFactory;
 import io.druid.segment.data.ByteBufferSerializer;
+import io.druid.segment.data.CompressedVSizeIntsIndexedSupplier;
 import io.druid.segment.data.GenericIndexed;
+import io.druid.segment.data.IndexedInts;
+import io.druid.segment.data.IndexedMultivalue;
 import io.druid.segment.data.IndexedRTree;
 import io.druid.segment.data.VSizeIndexed;
 import io.druid.segment.data.VSizeIndexedInts;
+import io.druid.segment.data.WritableSupplier;
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
 import java.nio.channels.WritableByteChannel;
 
-/**
- */
+
 public class DictionaryEncodedColumnPartSerde implements ColumnPartSerde
 {
-  private final boolean isSingleValued;
+  private static final int NO_FLAGS = 0;
+
+  enum Feature {
+    MULTI_VALUE;
+    public boolean isSet(int flags) { return (getMask() & flags) != 0; }
+    public int getMask() { return (1 << ordinal()); }
+  }
+
+  enum VERSION
```

Comment:
I would prefer something like `LEGACY_MULTI_VALUE_INDICATOR` and `LEGACY_SINGLE_VALUE_INDICATOR` but would be ok if it were just indicated more explicitly in comments.